internal/controller/operator/factory/vmdistributedcluster/vmagent.go

```diff
@@ -0,0 +1,298 @@
+		if k8serrors.IsNotFound(err) {
+			vmAgentExists = false
+			// If it doesn't exist, initialize object for creation.
+			vmAgentObj = &vmv1beta1.VMAgent{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      cr.Spec.VMAgent.Name,
+					Namespace: cr.Namespace,
+				},
+			}
+		} else {
+			return nil, fmt.Errorf("failed to get VMAgent %s/%s: %w", cr.Namespace, cr.Spec.VMAgent.Name, err)
+		}
+	}
+
+	// Prepare the desired spec for VMAgent.
+	desiredVMAgentSpec := vmv1alpha1.CustomVMAgentSpec{}
+	if cr.Spec.VMAgent.Spec != nil {
+		desiredVMAgentSpec = *cr.Spec.VMAgent.Spec.DeepCopy()
+	}
+
+	// Determine tenant id (default to "0" when not provided)
+	tenantID := "0"
+	if cr.Spec.VMAgent.TenantID != nil && *cr.Spec.VMAgent.TenantID != "" {
+		tenantID = *cr.Spec.VMAgent.TenantID
+	}
+
+	tenantPtr := &tenantID
+
+	// Point VMAgent to all VMClusters by constructing RemoteWrite entries
+	if len(vmClusters) > 0 {
+		desiredVMAgentSpec.RemoteWrite = make([]vmv1alpha1.CustomVMAgentRemoteWriteSpec, len(vmClusters))
+		for i, vmCluster := range vmClusters {
+			desiredVMAgentSpec.RemoteWrite[i].URL = remoteWriteURL(vmCluster, tenantPtr)
+		}
+	}
+
+	// Assemble new VMAgentSpec
+	newVMAgentSpec := vmv1beta1.VMAgentSpec{}
+	newVMAgentSpec.PodMetadata = desiredVMAgentSpec.PodMetadata
+	newVMAgentSpec.ManagedMetadata = desiredVMAgentSpec.ManagedMetadata
+	newVMAgentSpec.LogLevel = desiredVMAgentSpec.LogLevel
+	newVMAgentSpec.LogFormat = desiredVMAgentSpec.LogFormat
+	newVMAgentSpec.RemoteWriteSettings = desiredVMAgentSpec.RemoteWriteSettings
+	newVMAgentSpec.ShardCount = desiredVMAgentSpec.ShardCount
+	newVMAgentSpec.UpdateStrategy = desiredVMAgentSpec.UpdateStrategy
+	newVMAgentSpec.RollingUpdate = desiredVMAgentSpec.RollingUpdate
+	newVMAgentSpec.PodDisruptionBudget = desiredVMAgentSpec.PodDisruptionBudget
+	newVMAgentSpec.EmbeddedProbes = desiredVMAgentSpec.EmbeddedProbes
+	newVMAgentSpec.DaemonSetMode = desiredVMAgentSpec.DaemonSetMode
+	newVMAgentSpec.StatefulMode = desiredVMAgentSpec.StatefulMode
+	newVMAgentSpec.StatefulStorage = desiredVMAgentSpec.StatefulStorage
+	newVMAgentSpec.StatefulRollingUpdateStrategy = desiredVMAgentSpec.StatefulRollingUpdateStrategy
+	newVMAgentSpec.PersistentVolumeClaimRetentionPolicy = desiredVMAgentSpec.PersistentVolumeClaimRetentionPolicy
+	newVMAgentSpec.ClaimTemplates = desiredVMAgentSpec.ClaimTemplates
+	newVMAgentSpec.License = desiredVMAgentSpec.License
+	newVMAgentSpec.ServiceAccountName = desiredVMAgentSpec.ServiceAccountName
+	newVMAgentSpec.VMAgentSecurityEnforcements = desiredVMAgentSpec.VMAgentSecurityEnforcements
+	newVMAgentSpec.CommonDefaultableParams = desiredVMAgentSpec.CommonDefaultableParams
+	newVMAgentSpec.CommonConfigReloaderParams = desiredVMAgentSpec.CommonConfigReloaderParams
+	newVMAgentSpec.CommonApplicationDeploymentParams = desiredVMAgentSpec.CommonApplicationDeploymentParams
+
+	if len(desiredVMAgentSpec.RemoteWrite) > 0 {
```

Comment:
Consider to apply sorting to the vmagent.spec.remoteWrite. Otherwise code above may sort vmclusters differently and erase on disk buffer for vmagent due to url position change at args.